pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,486 @@
+	if sel.Empty() {
+		return []*api.Pod{}, nil
+	}
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	podList, err := dc.podLister.Pods(pdb.Namespace).List(sel)
+	if err != nil {
+		return []*api.Pod{}, err
+	}
+	pods := []*api.Pod{}
+	for _, p := range podList.Items {
+		obj, err := api.Scheme.DeepCopy(p)
+		if err != nil {
+			return []*api.Pod{}, err
+		}
+		pod := obj.(api.Pod)
+		pods = append(pods, &pod)
+	}
+	return pods, nil
+}
+
+func (dc *DisruptionController) worker() {
+	for {
+		func() {
+			key, quit := dc.queue.Get()
+			if quit {
+				return
+			}
+			defer dc.queue.Done(key)
+			glog.Infof("Syncing DB %q", key.(string))
+			if err := dc.sync(key.(string)); err != nil {
+				glog.Errorf("Error syncing PodDisruptionBudget %v, requeuing: %v", key.(string), err)
+				// TODO(mml): In order to be safe in the face of a total inability to write state
+				// changes, we should write an expiration timestamp here and consumers
+				// of the PDB state (the /evict subresource handler) should check that
+				// any 'true' state is relatively fresh.
+
+				// TODO(mml): file an issue to that effect
+				dc.queue.Add(key)
+			}
+		}()
+	}
+}
+
+func (dc *DisruptionController) sync(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing PodDisruptionBudget %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	obj, exists, err := dc.pdbLister.Store.GetByKey(key)
+	if !exists {
+		return err
+	}
+	if err != nil {
+		glog.Errorf("Unable to retrieve PodDisruptionBudget %v from store: %v", key, err)
+		return err
+	}
+
+	pdb := obj.(*policy.PodDisruptionBudget)
+
+	if err := dc.trySync(pdb); err != nil {
+		return dc.failSafe(pdb)
+	}
+
+	return nil
+}
+
+func (dc *DisruptionController) trySync(pdb *policy.PodDisruptionBudget) error {
+	pods, err := dc.getPodsForPdb(pdb)
+	if err != nil {
+		return err
+	}
+
+	expectedCount, desiredHealthy, err := dc.getExpectedPodCount(pdb, pods)
+	if err != nil {
+		return err
+	}
+
+	currentHealthy := dc.countHealthyPods(pdb, pods)
+	err = dc.updatePdbSpec(pdb, currentHealthy, desiredHealthy, expectedCount)
+
+	return err
+}
+
+func (dc *DisruptionController) getExpectedPodCount(pdb *policy.PodDisruptionBudget, pods []*api.Pod) (expectedCount, desiredHealthy int32, err error) {
+	err = nil
+	if pdb.Spec.MinAvailable.Type == intstr.Int {
+		desiredHealthy = pdb.Spec.MinAvailable.IntVal
+		expectedCount = int32(len(pods))
+	} else if pdb.Spec.MinAvailable.Type == intstr.String {
+		// A mapping from controllers to their scale.
+		controllerScale := map[types.UID]int32{}
+
+		// 1. Find the controller(s) for each pod.  If any pod has 0 controllers,
+		// that's an error.
+		for _, pod := range pods {
+			foundController := false
+			for _, finder := range dc.finders() {
+				var controllers []controllerAndScale
+				controllers, err = finder(pod)
+				if err != nil {
+					return
+				}
+				if len(controllers) > 0 {
```

Comment:
We actually *do* define what to do in this case, (`2. Add up all the controllers`).  Should I make this more clear?